File: src/modules/BodyProfileModule.cpp
#include "BodyProfileModule.h"

#include <QComboBox>
#include <QGroupBox>
#include <QLabel>
#include <QObject>
#include <QVBoxLayout>
#include <QWidget>

BodyProfileModule::BodyProfileModule() = default;

QString BodyProfileModule::id() const {
    return "body-profile";
}

QString BodyProfileModule::displayName() const {
    return "Body Profile";
}

QWidget* BodyProfileModule::view() {
    if (m_root == nullptr) {
        buildView();
    }

    return m_root;
}

void BodyProfileModule::initialize() {
    view();
    updateDescription(0);
}

void BodyProfileModule::buildView() {
    m_root = new QWidget();

    auto* layout = new QVBoxLayout(m_root);
    auto* intro = new QLabel("Simple sample module for choosing the base body profile.", m_root);
    intro->setWordWrap(true);

    auto* selectionBox = new QGroupBox("Body type", m_root);
    auto* selectionLayout = new QVBoxLayout(selectionBox);

    m_bodyType = new QComboBox(selectionBox);
    m_bodyType->addItems({"Male", "Female", "Neutral"});

    m_description = new QLabel(selectionBox);
    m_description->setWordWrap(true);

    selectionLayout->addWidget(m_bodyType);
    selectionLayout->addWidget(m_description);

    layout->addWidget(intro);
    layout->addWidget(selectionBox);
    layout->addStretch();

    QObject::connect(m_bodyType, qOverload<int>(&QComboBox::currentIndexChanged), m_root, [this](int index) {
        updateDescription(index);
    });
}

void BodyProfileModule::updateDescription(int index) {
    if (m_description == nullptr) {
        return;
    }

    switch (index) {
    case 0:
        m_description->setText("Base preset for an adult male body model.");
        break;
    case 1:
        m_description->setText("Base preset for an adult female body model.");
        break;
    default:
        m_description->setText("Neutral preset for early prototyping and shared testing.");
        break;
    }
}

File: src/modules/body_profile_module.py
from __future__ import annotations

from typing import Optional

from PySide6.QtWidgets import (
    QComboBox,
    QGroupBox,
    QLabel,
    QVBoxLayout,
    QWidget,
)

from core.imodule import IModule


class BodyProfileModule(IModule):
    def __init__(self) -> None:
        self._root: Optional[QWidget] = None
        self._body_type: Optional[QComboBox] = None
        self._description: Optional[QLabel] = None

    def id(self) -> str:
        return "body-profile"

    def display_name(self) -> str:
        return "Body Profile"

    def view(self) -> QWidget:
        if self._root is None:
            self._build_view()
        return self._root

    def initialize(self) -> None:
        self.view()
        self._update_description(0)

    def _build_view(self) -> None:
        self._root = QWidget()

        layout = QVBoxLayout(self._root)
        intro = QLabel(
            "Simple sample module for choosing the base body profile.",
            self._root,
        )
        intro.setWordWrap(True)

        selection_box = QGroupBox("Body type", self._root)
        selection_layout = QVBoxLayout(selection_box)

        self._body_type = QComboBox(selection_box)
        self._body_type.addItems(["Male", "Female", "Neutral"])

        self._description = QLabel(selection_box)
        self._description.setWordWrap(True)

        selection_layout.addWidget(self._body_type)
        selection_layout.addWidget(self._description)

        layout.addWidget(intro)
        layout.addWidget(selection_box)
        layout.addStretch()

        self._body_type.currentIndexChanged.connect(self._update_description)

    def _update_description(self, index: int) -> None:
        if self._description is None:
            return

        if index == 0:
            self._description.setText("Base preset for an adult male body model.")
            return

        if index == 1:
            self._description.setText("Base preset for an adult female body model.")
            return

        self._description.setText(
            "Neutral preset for early prototyping and shared testing."
        )